Sources/AngouriMath/Core/Entity/Omni/Entity.Matrix.cs

```diff
@@ -194,35 +194,32 @@ public Entity AsScalar()
             /// <summary>
             /// Finds the symbolical determinant via Laplace's method
             /// </summary>
-            public Entity Determinant => determinant.GetValue(
-                static @this => @this.InnerMatrix.DeterminantGaussianSafeDivision().InnerSimplified,
+            public Entity? Determinant => determinant.GetValue(
+                static @this => 
+                {
+                    if (!@this.IsSquare) 
+                        return null;
+                    return @this.InnerMatrix.DeterminantGaussianSafeDivision().InnerSimplified;
+                },
                 this
                 );
-            private FieldCache<Entity> determinant;
+            private FieldCache<Entity?> determinant;
 
-            // The reason it's not cached is because it throws exceptions.
-            /// <summary>Inverts the matrix</summary>
-            public Matrix ComputeInverse()
+            /// <summary>Returns an inverse matrix if it exists</summary>
+            public Matrix? Inverse => inverse.GetValue(static @this =>
             {
-                var cp = InnerMatrix.Copy(false);
-                try
-                {
-                    cp.InvertMatrix();
-                }
-                catch (InvalidShapeException)
-                {
-                    throw new InvalidMatrixOperationException("Cannot inverse a non-square matrix!");
-                }
-                catch (InvalidDeterminantException)
-                {
-                    throw new InvalidMatrixOperationException("Cannot inverse a singular matrix!");
-                }
+                var cp = @this.InnerMatrix.Copy(false);
+                if (!@this.IsSquare)
+                    return null;
+                if (@this.Determinant is null)
+                    return null;
+                if (@this.Determinant == 0)
```

Comment:
Shouldn't these checks be in GenericTensor?

Reply:
Same thing, we want to return null to avoid an exception

Reply:
Then don't throw exceptions in GenericTensor!

Reply:
Hmmmmmmmmmmmmmmmmmmmmmmmmmmm. We can have `1.0.1-noexcept` package for it. Since if you remember, there's `ALLOW_EXCEPTION` flag.

Reply:
@MomoDeve would love this